uqcsbot/dominos_coupons.py

```diff
@@ -0,0 +1,153 @@
+        keywords="Words to search for within the coupon. All coupons descriptions will mention at least one keyword.",
+    )
+    async def dominoscoupons(
+        self,
+        interaction: discord.Interaction,
+        number_of_coupons: Optional[int] = 5,
+        ignore_expiry: Optional[bool] = True,
+        keywords: Optional[str] = "",
+    ):
+        """
+        Returns a list of dominos coupons
+        """
+        await interaction.response.defer(thinking=True)
+
+        if number_of_coupons < 1 or number_of_coupons > MAX_COUPONS:
+            await interaction.edit_original_response(
+                content=f"You can't have that many coupons. Try a number between 1 and {MAX_COUPONS}."
+            )
+            return
+        try:
+            coupons = _get_coupons(number_of_coupons, ignore_expiry, keywords.split())
+        except RequestException as error:
+            logging.warning(
+                f"Could not connect to dominos coupon site ({COUPONESE_DOMINOS_URL}): {error.response.content}"
+            )
+            await interaction.edit_original_response(
+                content=f"Sadly could not reach the coupon website (<{COUPONESE_DOMINOS_URL}>)..."
+            )
+            return
+        except HTTPResponseException as error:
+            logging.warning(
+                f"Received a HTTP response code that was not OK (200), namely ({error.http_code}). Error information: {error.message}"
+            )
+            await interaction.edit_original_response(
+                content=f"Could not find the coupons on the coupon website (<{COUPONESE_DOMINOS_URL}>)..."
+            )
+            return
+
+        if not coupons:
+            await interaction.edit_original_response(
+                content=f"Could not find any coupons matching the given arguments from the coupon website (<{COUPONESE_DOMINOS_URL}>)."
+            )
+            return
+
+        message = "Domino's Coupons:\n"
+        for coupon in coupons:
+            message += f"`{coupon.code}` - {coupon.description} *[Expires: {coupon.expiry_date}]*\n"
+        await interaction.edit_original_response(content=message)
+
+
+class Coupon:
+    def __init__(self, code: str, expiry_date: str, description: str) -> None:
+        self.code = code
+        self.expiry_date = expiry_date
+        self.description = description
+
+    def is_valid(self) -> bool:
+        try:
+            expiry_date = datetime.strptime(self.expiry_date, "%Y-%m-%d")
+            now = datetime.now()
+            return all(
+                [
+                    expiry_date.year >= now.year,
+                    expiry_date.month >= now.month,
+                    expiry_date.day >= now.day,
+                ]
+            )
+        except ValueError:
+            return True
+
+    def keyword_matches(self, keyword: str) -> bool:
+        return keyword.lower() in self.description.lower()
+
+
+class HTTPResponseException(Exception):
```

Comment:
Personal preference: putting exception types at the top of the file keeps them all in one spot and highlights them for anyone who's quickly skimming.